player/loadfile.c

```diff
@@ -1208,13 +1209,16 @@ static void open_demux_reentrant(struct MPContext *mpctx)
         bool failed = done && !mpctx->open_res_demuxer;
         bool correct_url = strcmp(mpctx->open_url, url) == 0;
 
-        if (correct_url && !failed) {
+        if (correct_url && !mpctx->demuxer_changed && !failed) {
             MP_VERBOSE(mpctx, "Using prefetched/prefetching URL.\n");
-        } else if (correct_url && failed) {
-            MP_VERBOSE(mpctx, "Prefetched URL failed, retrying.\n");
-            cancel_open(mpctx);
         } else {
-            if (done) {
+            if (correct_url && !mpctx->demuxer_changed && failed) {
```

Comment:
It is hard to read for me. Not sure I get it right, but I would try to reduce `demuxer_changed` checks,

``` c
if (correct_url && !mpctx->demuxer_changed && !failed) {
    MP_VERBOSE(mpctx, "Using prefetched/prefetching URL.\n");
} else {
    if (correct_url && failed) {
        MP_VERBOSE(mpctx, "Prefetched URL failed, retrying.\n");
    } else if (mpctx->demuxer_changed) {
        if (done) {
            MP_VERBOSE(mpctx, "Dropping finished prefetch because demuxer options changed.\n");
        } else {
            MP_VERBOSE(mpctx, "Aborting ongoing prefetch because demuxer options changed.\n");
        }
    } else {
        if (done) {
            MP_VERBOSE(mpctx, "Dropping finished prefetch of wrong URL.\n");
        } else {
            MP_VERBOSE(mpctx, "Aborting ongoing prefetch of wrong URL.\n");
        }
    }
    cancel_open(mpctx);
}
```